### PR TITLE
New module: minion.list

### PR DESCRIPTION
**Pull request by andygrunwald**
I described a possible feature in https://github.com/saltstack/salt/issues/28663.

This Pull Requests adds a new module to list all minions.
Of course, this module makes only sense to execute on a master, because a typical minion don't know about other minions.
This module is (nearly) the same as `salt-key -L`. But as a module it is available by REST-API. `salt-key` is only available as a shell command.

This is more like a draft, but maybe some people who are more familiar with the saltstack codebase can help to answer some questions (i will be happy to improve the code):
- Is there a smoother way to get the information about all minions? Is this information only stored in the filesystem at the `pki_dir` of the master?
- Is it possible to detect if the (current) minion is a master? With this we can skip the I/O on the minion
- This works only if the salt-master runs a salt-minion itself. Any other idea?
- Do you think this is a good idea?
- Do you suggest a better way to do this in general?

Call:

```
$ salt -l debug -c ./etc/salt 'saltdev' minion.list
```

Answer:

```
[DEBUG   ] Reading configuration from /private/tmp/salt-env/etc/salt/master
[DEBUG   ] Missing configuration file: /Users/agrunwald/.saltrc
[DEBUG   ] Configuration file path: /private/tmp/salt-env/etc/salt/master
[WARNING ] Insecure logging configuration detected! Sensitive data may be logged.
[DEBUG   ] Reading configuration from /private/tmp/salt-env/etc/salt/master
[DEBUG   ] Missing configuration file: /Users/agrunwald/.saltrc
[DEBUG   ] MasterEvent PUB socket URI: ipc:///tmp/salt-env/var/run/salt/master/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///tmp/salt-env/var/run/salt/master/master_event_pull.ipc
[DEBUG   ] Initializing new AsyncZeroMQReqChannel for ('/tmp/salt-env/etc/salt/pki/master', '172.25.10.26_master', 'tcp://127.0.0.1:4506', 'clear')
[DEBUG   ] LazyLoaded local_cache.get_load
[DEBUG   ] get_iter_returns for jid 20151106222329855220 sent to set(['saltdev']) will timeout at 22:23:34.860702
[DEBUG   ] jid 20151106222329855220 return from saltdev
[DEBUG   ] LazyLoaded nested.output
saltdev:
    ----------
    minions:
        - saltdev
    minions_denied:
    minions_pre:
    minions_rejected:
[DEBUG   ] jid 20151106222329855220 found all minions set(['saltdev'])
```

### CODE REVIEW

File: salt/modules/minion.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    Return a list of accepted, denied, unaccepted and rejected keys.
+    This is the same output as `salt-key -L`
+
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt 'master' minion.list
+    '''
+    pki_dir = globals().get('__salt__')['config.get']('pki_dir', '')
```
Comment (cachedout) [inline salt/modules/minion.py]: The loader automatically injects `__salt__` into the module's namespace. Instead, this should be simply `__salt__['config.get']('pki_dir', '')`

**Comment by cachedout** [comment]
Hi @andygrunwald. I left some comments in-line on this PR with some suggestions. Also, it does look like there is a lint error that needs to be taken care of before we can merge this. Could you please have a look and let me know what you think? Thanks!

**Comment by andygrunwald** [comment]
@cachedout Thanks. I will fix it and test it again. Where i can see details about the lint error?

**Comment by cachedout** [comment]
Hi @andygrunwald Thanks for looking at this!

The pylint error is a little obfuscated I'm afraid. I've pasted it below. You just need to add one line:

```
************* Module salt.modules.minion
minion.py:5: [W1699(incompatible-py3-code), ] Incompatible Python 3 code found. Proposed fix:
 '''
 Module to provide information about minions
 '''
+from __future__ import absolute_import
 
 # Import Python libs
 import os
```

**Comment by lorengordon** [comment]
Would it be desirable to rename this function, to avoid a name collision with the native python function `list()`?

**Comment by cachedout** [comment]
@lorengordon Great suggestion!

@andygrunwald See the `cpan` salt module for an example of doing this with the `list` function there. Note the `__func_alias__` block near the top of the file.

**Comment by andygrunwald** [comment]
Thanks for your feedback.
I fixed the Pylint error, used the global `__salt__` variable and implemented the function alias to avoid conflicts with `list()`.
To be honest: I did the first feedback round here (changes) without testing. Before merging this should be  tested again. The original PR was tested.

Like to get new feedback :)

File: salt/modules/minion.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+    CLI Example:
+
+    .. code-block:: bash
+
+        salt 'master' minion.list
+    '''
+    pki_dir = __salt__['config.get']('pki_dir', '')
+    transport = __salt__['config.get']('transport', '')
+
+    # We have to replace the minion/master directoryies
```
Comment (cachedout) [inline salt/modules/minion.py]: `directories` and not `directoryies` please

File: salt/modules/minion.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+    pki_dir = pki_dir.replace("minion", "master")
```
Comment (cachedout) [inline salt/modules/minion.py]: Single-quotes here please

**Comment by cachedout** [comment]
Hi @andygrunwald. This looks really good. I left two comments on minor issues but after you take a look at them I think that we should be ready to get this merged. Thanks!

**Comment by andygrunwald** [comment]
@cachedout Updated according your comments.

**Comment by cachedout** [comment]
@andygrunwald Looks great! Thank you for this. Many people will find it useful.

**Pull request merged**